### ISSUE TITLE
Registering new tasks after the Scheduler has been created

### ISSUE THREAD

**Issue by Desislav-Petrov** [kagkarlsson/db-scheduler#220]
Hi,

I've had an interesting use case recently where I needed to execute one off tasks in db scheduler and use its ability to ensure only 1 instance is executing the task.

The problem is that I don't know all the tasks ahead of creating the scheduler so I think it would be useful if you can 'register' tasks even after the scheduler has been created.

What do people think about this feature? If it's considered useful, I can put up a PR. 

Thanks

**Comment by kagkarlsson** [comment]
This is a common misconception about db-scheduler. The `Scheduler` obviously needs to know about the _implementation_ on startup. But you can have as many _instances_ of a certain `Task` as you like, and these can be added dynamically.

**Comment by Desislav-Petrov** [comment]
Yes, that makes sense, however the use case I'm talking about is different.
1) Db scheduler is created with a certain number of tasks
2) After that, we'd like to add a brand new Task definition (which hasn't been supplied during the creation of the schedulre) and create instances of it at a later point

This is impossible at the moment because, as you said, you have to supply the definitions on startup. So what I'm saying is, let's add a feature that allows you to load new task definitions post the creation of the scheduler, which will then allow you to create instances of the new task.

Does this make sense?

**Comment by kagkarlsson** [comment]
I am missing something.

> After that, we'd like to add a brand new Task definition (which hasn't been supplied during the creation of the schedulre) and create instances of it at a later point

You can supply task definitions (`Task`) at startup without creating any instances of it

**Comment by kagkarlsson** [comment]
(Is this a recurring task that you want to start at a later time?)

**Comment by Desislav-Petrov** [comment]
> I am missing something.
> 
> > After that, we'd like to add a brand new Task definition (which hasn't been supplied during the creation of the schedulre) and create instances of it at a later point
> 
> You can supply task definitions (`Task`) at startup without creating any instances of it

Apologies, I might not have explained it clearly enough. Let me try again.

The problem is that sometimes you will only know the exact definition of a task after the scheduler has already been created. In situations like this, your only option is to destroy & recreate the scheduler to add all old definitions + the new definition.

The idea of the feature is to allow us, at runtime, post the creation of the scheduler to add new task definitions.

Does this make more sense?

**Comment by kagkarlsson** [comment]
I think I am starting to see where you are going. You are dependent on some data that you receive at runtime? For such a case, you need to create a `Task<YourData>` that is seeded with the data when the task-instance is created. The data can then be used by the `ExecutionHandler` and your code do different things based on it.

You suggestion on adding `Task` dynamically has a big problem. You will not be able to add it to all instances of the `Scheduler` and the other instances will fail on execution since they do not know the task

**Comment by Desislav-Petrov** [comment]
Yes, that's the use case. I thought about that but it didn't look very clean.. Can you think of any other options?

And yes, i see your point about breaking the cluster functionality with that feature. Unless you start storing task definitions in a serialized format in the database for example but that opens another can of worms..

**Comment by kagkarlsson** [comment]
> Can you think of any other options?

Not really

> Unless you start storing task definitions in a serialized format in the database for example but that opens another can of worms..

Yeah, I want to avoid that type of complexity. To be honest, that sounds "less clean" than just serializing the data which we do now :)

**Comment by Desislav-Petrov** [comment]
Ok cool.. looks like we don't have a clean solution to this problem apart from adding complexity but given that nobody else apart from me has mentioned this it's probably too much effort for too little value. Shall we close the issue in that case?
